cmd/kubeadm/app/phases/controlplane/manifests_linux.go

```diff
@@ -0,0 +1,151 @@
+// +build linux
+
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controlplane
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	certphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/certs"
+	"k8s.io/kubernetes/cmd/kubeadm/app/util/users"
+	"k8s.io/utils/pointer"
+)
+
+type pathOwerAndPermissionsUpdaterFunc func(path string, uid, gid int64, perms uint32) error
+
+// runComponentAsNonRoot updates the pod manifest and the hostVolume permissions to run as non root.
+func runComponentAsNonRoot(componentName string, pod *v1.Pod, usersAndGroups *users.UsersAndGroups, cfg *kubeadmapi.ClusterConfiguration) error {
+	switch componentName {
+	case kubeadmconstants.KubeAPIServer:
+		return runKubeAPIServerAsNonRoot(pod, usersAndGroups, cfg, users.UpdatePathOwnerAndPermissions)
```

Comment:
individual component helper functions should accept prepared `uid, gid, .... int64` instead of the `UsersAndGroups` object. this would avoid making Entry in `users_linux.go` public due to unit testing needs.

[1]

Reply:
Done.